File: src/org/elixir_lang/beam/Beam.java
package org.elixir_lang.beam;

import com.ericsson.otp.erlang.OtpErlangDecodeException;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.util.indexing.FileContent;
import gnu.trove.THashMap;
import org.elixir_lang.beam.chunk.Atoms;
import org.elixir_lang.beam.chunk.CallDefinitions;
import org.elixir_lang.beam.chunk.Chunk;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.ByteArrayInputStream;
import java.io.DataInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.Charset;
import java.util.*;
import java.util.stream.Stream;
import java.util.zip.GZIPInputStream;

import static org.elixir_lang.beam.chunk.Chunk.TypeID.*;
import static org.elixir_lang.beam.chunk.Chunk.length;
import static org.elixir_lang.beam.chunk.Chunk.typeID;

/**
 * See http://beam-wisdoms.clau.se/en/latest/indepth-beam-file.html
 */
public class Beam {
    private static final List<Chunk.TypeID> CALL_DEFINITION_TYPE_IDS = Arrays.asList(EXPT, LOCT);

    private static final int GZIP_FIRST_UNSIGNED_BYTE = 0x1f;
    private static final int GZIP_SECOND_UNSIGNED_BYTE = 0x8b;

    private static final String HEADER = "FOR1";

    /*
     * CONSTANTS
     */
    private static final Logger LOGGER = Logger.getInstance(Beam.class);

    /*
     * Fields
     */
    private Map<String, Chunk> chunkByTypeID;

    /*
     * Constructors
     */

    private Beam(@NotNull Collection<Chunk> chunkCollection) {
        chunkByTypeID = new THashMap<String, Chunk>(chunkCollection.size());

        for (Chunk chunk : chunkCollection) {
            chunkByTypeID.put(chunk.typeID, chunk);
        }
    }

    /*
     * Static Methods
     */

    @Nullable
    public static Beam from(@NotNull DataInputStream dataInputStream, @NotNull String path) {
        String header;

        try {
            header = typeID(dataInputStream, path);
        } catch (IOException ioException) {
            LOGGER.error("Could not read header from BEAM DataInputStream from " + path, ioException);
            return null;
        }

        if (!HEADER.equals(header)) {
            LOGGER.error("header typeID (" + header + ") did not match expected (" + HEADER + ") from " + path);
            return null;
        }

        try {
            length(dataInputStream);
        } catch (IOException ioException) {
            LOGGER.error("Could not read length from BEAM DataInputStream from " + path, ioException);
            return null;
        }

        String section;

        try {
            section = typeID(dataInputStream, path);
        } catch (IOException ioException) {
            LOGGER.error("Could not read section header from BEAM DataInputStream from " + path, ioException);
            return null;
        }


        if (!"BEAM".equals(section)) {
            LOGGER.error("Section header is not BEAM in " + path);
            return null;
        }

        List<Chunk> chunkList = new ArrayList<Chunk>();
        int i = 1;

        while (true) {
            Chunk chunk;

            try {
                chunk = Chunk.from(dataInputStream, path);
            } catch (IOException ioException) {
                LOGGER.error(
                        "Could not read chunk number " + i + " from BEAM DataInputStream from " + path +
                                ".  Returning truncated Beam object",
                        ioException
                );
                break;
            }

            if (chunk != null) {
                chunkList.add(chunk);
            } else {
                break;
            }
        }

        return new Beam(chunkList);
    }

    @NotNull
    public static Optional<Beam> from(@NotNull byte[] content, @NotNull String path)
            throws IOException, OtpErlangDecodeException {
        return decompressedInputStream(new ByteArrayInputStream(content))
                .map(DataInputStream::new)
                .flatMap(dataInputStream -> Optional.ofNullable(Beam.from(dataInputStream, path)));
    }

    @NotNull
    public static Optional<Beam> from(@NotNull FileContent fileContent) throws IOException, OtpErlangDecodeException {
        return from(fileContent.getContent(), fileContent.getFile().getPath());
    }

    private static Optional<InputStream> virtualFileToInputStream(@NotNull VirtualFile virtualFile) {
        Optional<InputStream> inputStreamOptional;

        try {
            inputStreamOptional = Optional.of(virtualFile.getInputStream());
        } catch (IOException e) {
            inputStreamOptional = Optional.empty();
        }

        return inputStreamOptional;
    }

    private static Optional<InputStream> decompressedInputStream(@NotNull InputStream inputStream) {
        assert inputStream.markSupported();

        inputStream.mark(2);
        Optional<InputStream> decompressedInputStreamOptional;

        try {
            if (inputStream.read() == GZIP_FIRST_UNSIGNED_BYTE && inputStream.read() == GZIP_SECOND_UNSIGNED_BYTE) {
                inputStream.reset();
                decompressedInputStreamOptional = Optional.of(new GZIPInputStream(inputStream));
            } else {
                inputStream.reset();
                decompressedInputStreamOptional = Optional.of(inputStream);
            }
        } catch (IOException e) {
            decompressedInputStreamOptional = Optional.empty();
        }

        return decompressedInputStreamOptional;
    }

    @NotNull
    public static Optional<Beam> from(@NotNull VirtualFile virtualFile) {
        return virtualFileToInputStream(virtualFile)
                .flatMap(Beam::decompressedInputStream)
                .map(DataInputStream::new)
                .flatMap(dataInputStream -> Optional.ofNullable(Beam.from(dataInputStream, virtualFile.getPath())));
    }

    public static boolean is(@NotNull VirtualFile virtualFile) {
        return !virtualFile.isDirectory() && "beam".equals(virtualFile.getExtension());
    }

    /*
     * Instance Methods
     */

    @Nullable
    public Atoms atoms() {
        Atoms atoms = null;

        Chunk chunk = chunk(ATOM);

        if (chunk != null) {
            atoms = Atoms.from(chunk, ATOM, Charset.forName("LATIN1"));
        } else {
            chunk = chunk(ATU8);

            if (chunk != null) {
                atoms = Atoms.from(chunk, ATU8, Charset.forName("UTF-8"));
            }
        }

        return atoms;
    }

    @Nullable
    private Chunk chunk(@NotNull String typeID) {
        return chunkByTypeID.get(typeID);
    }

    @Nullable
    private Chunk chunk(@NotNull Chunk.TypeID typeID) {
        return chunk(typeID.toString());
    }

    @NotNull
    private Optional<CallDefinitions> callDefinitions(@NotNull Chunk.TypeID typeID) {
        Optional<CallDefinitions> callDefinitions;

        Chunk chunk = chunk(typeID);

        if (chunk != null) {
            callDefinitions = Optional.ofNullable(CallDefinitions.from(chunk, typeID));
        } else {
            callDefinitions = Optional.empty();
        }

        return callDefinitions;
    }

    @NotNull
    public Stream<CallDefinitions> callDefinitionsStream() {
        return CALL_DEFINITION_TYPE_IDS
                .stream()
                .flatMap(typeID -> callDefinitions(typeID).map(Stream::of).orElseGet(Stream::empty));
    }
}

File: src/org/elixir_lang/beam/Beam.kt
package org.elixir_lang.beam

import com.ericsson.otp.erlang.OtpErlangDecodeException
import com.ericsson.otp.erlang.OtpErlangObject
import com.ericsson.otp.erlang.OtpInputStream
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.util.indexing.FileContent
import org.elixir_lang.beam.chunk.*
import org.elixir_lang.beam.chunk.Chunk.TypeID.*
import org.elixir_lang.beam.chunk.Chunk.length
import org.elixir_lang.beam.chunk.Chunk.typeID
import org.elixir_lang.beam.term.ByteCount
import java.io.*
import java.nio.charset.Charset
import java.util.*
import java.util.zip.GZIPInputStream

private val callDefinitionTypeIDs = arrayOf(EXPT, LOCT)

private const val GZIP_FIRST_UNSIGNED_BYTE = 0x1f
private const val GZIP_SECOND_UNSIGNED_BYTE = 0x8b
private const val HEADER = "FOR1"

fun binaryToTerm(byteArray: ByteArray, offset: Int): Pair<OtpErlangObject, ByteCount> {
    val stream = OtpInputStream(byteArray, offset, byteArray.size, 0)
    val term = stream.read_any()

    return Pair(term, stream.pos - offset)
}

private fun decompressedInputStream(inputStream: InputStream): InputStream? {
    assert(inputStream.markSupported())

    inputStream.mark(2)

    return try {
        if (inputStream.read() == GZIP_FIRST_UNSIGNED_BYTE && inputStream.read() == GZIP_SECOND_UNSIGNED_BYTE) {
            inputStream.reset()
            GZIPInputStream(inputStream)
        } else {
            inputStream.reset()
            inputStream
        }
    } catch (e: IOException) {
        null
    }
}

private fun virtualFileToInputStream(virtualFile: VirtualFile): InputStream? =
        try {
            virtualFile.inputStream
        } catch (e: IOException) {
            null
        }

/**
 * See http://beam-wisdoms.clau.se/en/latest/indepth-beam-file.html
 */
class Beam private constructor(chunkCollection: Collection<Chunk>) {
    private val chunkByTypeID: Map<String, Chunk> = chunkCollection.associateBy { it.typeID }

    fun atoms(): Atoms? =
            chunk(ATOM)?.let { Atoms.from(it, ATOM, Charset.forName("LATIN1")) } ?:
                    chunk(ATU8)?.let { Atoms.from(it, ATU8, Charset.forName("UTF-8")) }

    fun attributes(): Keyword? = chunk(ATTR)?.let(::from)
    private fun chunk(typeID: String): Chunk? = chunkByTypeID[typeID]
    private fun chunk(typeID: Chunk.TypeID): Chunk? = chunk(typeID.toString())
    fun chunkCollection(): Collection<Chunk> = chunkByTypeID.values

    private fun callDefinitions(typeID: Chunk.TypeID, atoms: Atoms?): CallDefinitions? =
        chunk(typeID)?.let {
            CallDefinitions.from(it, typeID, atoms)
        }

    fun callDefinitionsList(atoms: Atoms?): List<CallDefinitions> =
        callDefinitionTypeIDs.mapNotNull { typeID ->
            callDefinitions(typeID, atoms)
        }

    fun code(): Code? = chunk(CODE)?.let { Code.from(it) }
    fun compileInfo(): Keyword? = chunk(CINF)?.let(::from)
    fun elixirDocumentation(): ElixirDocumentation? = chunk(EXDC)?.let { ElixirDocumentation.from(it) }
    fun debugInfo(): DebugInfo? = chunk(DBGI)?.let { org.elixir_lang.beam.chunk.debug_info.from(it) }
    fun exports(atoms: Atoms?): CallDefinitions? = callDefinitions(EXPT, atoms)
    fun functions(atoms: Atoms?): Functions? = chunk(FUNT)?.let { Functions.from(it, atoms) }

    fun imports(atoms: Atoms?): Imports? =
        chunk(IMPT)?.let {
            Imports.from(it, atoms)
        }

    fun lines(): Lines? = chunk(LINE)?.let { Lines.from(it) }
    fun literals(): Literals? = chunk(LITT)?.let { Literals.from(it) }
    fun locals(atoms: Atoms?): CallDefinitions? = callDefinitions(LOCT, atoms)
    fun strings(): Strings? = chunk(STRT)?.let { Strings.from(it) }

    companion object {
        private val LOGGER = Logger.getInstance(Beam::class.java)

        fun from(dataInputStream: DataInputStream, path: String): Beam? {
            val header: String?

            try {
                header = typeID(dataInputStream, path)
            } catch (ioException: IOException) {
                LOGGER.error("Could not read header from BEAM DataInputStream from " + path, ioException)
                return null
            }

            if (HEADER != header) {
                if (header != null && !testCase(header)) {
                    LOGGER.error(
                            "header typeID ($header) did not match expected ($HEADER) from $path. " +
                                    "There are ${dataInputStream.available()} bytes available on the " +
                                    "dataInputStream. File size is ${File(path).length()} bytes."
                    )
                }
                return null
            }

            try {
                length(dataInputStream)
            } catch (ioException: IOException) {
                LOGGER.error("Could not read length from BEAM DataInputStream from " + path, ioException)
                return null
            }

            val section: String?

            try {
                section = typeID(dataInputStream, path)
            } catch (ioException: IOException) {
                LOGGER.error("Could not read section header from BEAM DataInputStream from " + path, ioException)
                return null
            }


            if ("BEAM" != section) {
                LOGGER.error("Section header is not BEAM in " + path)
                return null
            }

            val chunkList = ArrayList<Chunk>()
            val i = 1

            while (true) {
                val chunk: Chunk?

                try {
                    chunk = Chunk.from(dataInputStream, path)
                } catch (ioException: IOException) {
                    LOGGER.error(
                            "Could not read chunk number " + i + " from BEAM DataInputStream from " + path +
                                    ".  Returning truncated Beam object",
                            ioException
                    )
                    break
                }

                if (chunk != null) {
                    chunkList.add(chunk)
                } else {
                    break
                }
            }

            return Beam(chunkList)
        }

        @Throws(IOException::class, OtpErlangDecodeException::class)
        fun from(content: ByteArray, path: String): Beam? =
                decompressedInputStream(ByteArrayInputStream(content))
                        ?.let { DataInputStream(it) }
                        ?.let { Beam.from(it, path) }

        @Throws(IOException::class, OtpErlangDecodeException::class)
        fun from(fileContent: FileContent): Beam? = from(fileContent.content, fileContent.file.path)

        fun from(virtualFile: VirtualFile): Beam? =
                virtualFileToInputStream(virtualFile)
                        ?.let { decompressedInputStream(it) }
                        ?.let { DataInputStream(it) }
                        ?.let { Beam.from(it, virtualFile.path) }

        fun `is`(virtualFile: VirtualFile): Boolean = !virtualFile.isDirectory && "beam" == virtualFile.extension

        private fun testCase(header: String?): Boolean = header == "baz "
    }
}
